lib/encodings/base32.fz

```diff
@@ -0,0 +1,169 @@
+# This file is part of the Fuzion language implementation.
+#
+# The Fuzion language implementation is free software: you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as published
+# by the Free Software Foundation, version 3 of the License.
+#
+# The Fuzion language implementation is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
+# License for more details.
+#
+# You should have received a copy of the GNU General Public License along with The
+# Fuzion language implementation.  If not, see <https://www.gnu.org/licenses/>.
+
+
+# -----------------------------------------------------------------------
+#
+#  Tokiwa Software GmbH, Germany
+#
+#  Source code of Fuzion standard library feature base32
+#
+# -----------------------------------------------------------------------
+
+# Base32 encoding and decoding as defined in RFC 4648
+#
+public base32 is
+
+  # allows redefinition e.g. for base32hex
+  module get_alphabet => "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567".utf8.as_array
+  module encoding_name => "base32"
+
+  # NYI: compiler does currently not optimize features without arguments to treat them like fields
+  alphabet := get_alphabet
+
+  # decode a valid base32 characters to 5 bits
+  # visibility module to allow redefinition in base32hex
+  module quintet_bits(n u8) =>
+    if n >= 65 && n <= 90        # case A-Z
+      n.as_u64 - 65
+
+    else #if n >= 50 && n <= 55  # case 2-7
+      n.as_u64 - 24
+
+  # checks if a character is valid in the encoding
+  # visibility module to allow redefinition in base32hex
+  module is_valid(c u8) =>
+    (c >= 65 && c <= 90) || (c >= 50 && c <= 55)
+
+  # Encodes a given byte sequence in base32, output is padded to multiple of 8
+  # returns a sequence of ascii values
+  public encode(data) =>
+
+    # extract 8 quintets from 40 bit (of an 64 bit integer)
+    enc40(n u64) =>
+      block := array i32 8 (i->((n >> (35-i*5).as_u64) & 31).as_i32)
+      block.map (x->alphabet[x])
```

Comment:
does this work?
```
array u8 8 i->
   idx := (n.as_i32 >> 35-i*5) & 31
   alphabet[idx]
```

Reply:
cast to i32 needs to happen at the end, but aside from that yes 👍️